src/QsCompiler/CompilationManager/EditorCompletion.cs

```diff
@@ -0,0 +1,686 @@
+                completions =
+                    Keywords.ReservedKeywords
+                    .Select(keyword => new CompletionItem { Label = keyword, Kind = CompletionItemKind.Keyword })
+                    .Concat(GetLocalCompletions(file, compilation, position))
+                    .Concat(GetCallableCompletions(file, compilation, openNamespaces))
+                    .Concat(GetTypeCompletions(file, compilation, openNamespaces))
+                    .Concat(GetGlobalNamespaceCompletions(compilation))
+                    .Concat(GetNamespaceAliasCompletions(file, compilation, position));
+            }
+            else
+            {
+                completions = Array.Empty<CompletionItem>();
+            }
+            return new CompletionList { IsIncomplete = false, Items = completions.ToArray() };
+        }
+
+        /// <summary>
+        /// Updates the given completion item with additional information if any is available. The completion item
+        /// returned is a reference to the same completion item that was given; the given completion item is mutated
+        /// with the additional information.
+        /// <para/>
+        /// Returns null (and the item is not updated) if any argument is null.
+        /// </summary>
+        public static CompletionItem ResolveCompletion(
+            this CompilationUnit compilation, CompletionItem item, CompletionItemData data, MarkupKind format)
+        {
+            if (compilation == null || item == null || data == null)
+                return null;
+            var documentation = TryGetDocumentation(compilation, data, item.Kind, format == MarkupKind.Markdown);
+            if (documentation != null)
+                item.Documentation = new MarkupContent { Kind = format, Value = documentation };
+            return item;
+        }
+
+        /// <summary>
+        /// Returns the completion environment at the given position in the file or null if the environment cannot be
+        /// determined.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">Thrown when any argument is null.</exception>
+        private static CompletionEnvironment GetCompletionEnvironment(FileContentManager file, Position position)
+        {
+            if (file == null)
+                throw new ArgumentNullException(nameof(file));
+            if (position == null)
+                throw new ArgumentNullException(nameof(position));
+
+            var relationship = GetFragmentAtOrBefore(file, position, out var index, out _);
+            CodeFragment.TokenIndex parent = null;
+            if (relationship == FragmentRelationship.EnclosingScope)
+                parent = index;
+            else if (relationship == FragmentRelationship.Self || relationship == FragmentRelationship.SameScope)
+                parent = index.GetNonEmptyParent();
+            else if (relationship == FragmentRelationship.InnerScope)
```

Comment:
Wouldn't you have to count if you have several closing brackets?

Reply:
If there are multiple closing brackets in a row, e.g.

```
operation Foo () : Unit {
    if (true) {
        Bar();
    }
}
```

Then if the position is after the last closing bracket, the fragment found by `GetFragmentAtOrBefore` seems to be an empty fragment that starts and ends at the same position as the last bracket, and is followed by `}`. I'm not sure why this is the case - maybe I shouldn't rely on it...

Reply:
This code has been generalized to use relative indentation numbers now.